perception/sub8_perception/src/sub8_perception/start_gate.cpp

```diff
@@ -0,0 +1,204 @@
+  service_enable_ =
+      nh_.advertiseService("/vision/start_gate/enable", &Sub8StartGateDetector::requestStartGateEnable, this);
+}
+Sub8StartGateDetector::~Sub8StartGateDetector()
+{
+}
+
+void Sub8StartGateDetector::imageCallback(const sensor_msgs::ImageConstPtr &image_msg,
+                                           const sensor_msgs::CameraInfoConstPtr &info_msg)
+{
+  if (running_)
+  {
+    // Get some image/camera info
+    cam_model_.fromCameraInfo(info_msg);
+    image_time_ = image_msg->header.stamp;
+
+    // If we have a large enough sample of gate identifications:
+    if (gate_line_buffer_.size() > gate_line_buffer_.capacity() - 1)
+    {
+      // Loop through the buffer and save mean and deviations of both size and
+      // position
+      for (auto &gate1 : gate_line_buffer_)
+      {
+        for (auto &gate2 : gate1)
+        {
+          cv::Point2i s = gate2[1] - gate2[0];
+          cv::Point2i center = cv::Point2i(s.x / 2 + s.x, s.y / 2 + s.y);
+          accX_(center.x);
+          accY_(center.y);
+          accSizeX_(s.x);
+          accSizeY_(s.y);
+        }
+      }
+
+      // Look at one of the elements (gate_line_buffer_.capacity()-1) in the
+      // circular buffer and check it's statistics and delete if necessary
+      for (size_t i = 0; i < gate_line_buffer_[gate_line_buffer_.capacity() - 1].size(); i++)
+      {
+        cv::Point2i s = gate_line_buffer_[gate_line_buffer_.capacity() - 1][i][1] -
+                        gate_line_buffer_[gate_line_buffer_.capacity() - 1][i][0];
+        cv::Point2i center = cv::Point2i(s.x / 2 + s.x, s.y / 2 + s.y);
+
+        // If position of the element is greater than one deviation from the
+        // mean, then delete
+        if (abs(center.x - mean(accX_)) > sqrt(variance(accX_)) || abs(center.y - mean(accY_) > sqrt(variance(accY_))))
+        {
+          gate_line_buffer_[gate_line_buffer_.capacity() - 1].erase(
+              gate_line_buffer_[gate_line_buffer_.capacity() - 1].begin() + i);
+          i--;
+          std::cout << "Deleted: " << center << std::endl;
+          continue;
+        }
+
+        // If the size of the element is greater than one deviation from the
+        // mean, then delete
+        if (abs(s.x - mean(accSizeX_)) > sqrt(variance(accSizeX_)) ||
+            abs(s.y - mean(accSizeY_)) > sqrt(variance(accSizeY_)))
+        {
+          gate_line_buffer_[gate_line_buffer_.capacity() - 1].erase(
+              gate_line_buffer_[gate_line_buffer_.capacity() - 1].begin() + i);
+          i--;
+          std::cout << "Deleted size: " << s << std::endl;
+          continue;
+        }
+      }
+
+      // Visualize on of the saved gates in the buffer
+
+      // #ifdef VISUALIZE
+      cv::Mat show = cv_bridge::toCvShare(image_msg, "bgr8")->image;
+      for (auto &gate : gate_line_buffer_[29])
+      {
+        cv::rectangle(show, gate[0], gate[1], cv::Scalar(255, 0, 0), 2, 8, 0);
+      }
+      cv::imshow("ay", show);
```

Comment:
Instead of using defines for visualization, the cleaner way would be to publish debug images.